### PR TITLE
Improve hud accessibility for screen reader users

### PR DESCRIPTION
**Pull request by zersiax**
Signed-off-by: Florian Beijers <florianbeijers@gmail.com>

### CODE REVIEW

**Comment by zersiax** [comment]
This PR does a number of things:

- the various iframes are now titled, which makes their function and location clearer. This should help in making the tutorial easier to follow.
- The roles and labels of some of the components have been overwritten for screen reader users only. This will make identification of the various controls easier. Notably, buttons are actually buttons semantically now where they were divs before, and the labels for the various tools are easier to follow now.

Some of the labels are currently English-only, and these changes do nothing for keyboard-only users just yet, but I'm intending to add some more odds and ends as I think of them.

**Comment by thc202** [comment]
The formatting in `inject.js` needs to be fixed: e.g. `571:1 Expected indentation of 2 tabs but found 0.                                  indent`.

**Comment by thc202** [comment]
We can also addressed that, as you prefer.

**Comment by psiinon** [comment]
Thanks for submitting this - we're always very happy to see accessibility improvements.

It should be easy to make the labels I18n, eg change: 
`<modal :title="title" :show="show" @close="close" aria-label="Select Tools">`
to something like
`<modal :title="title" :show="show" @close="close" aria-label="$t('message.label_select_tools')">`
Then add the following line to src/other/resources/UIMessages/UIMessages.properties:
`label_select_tools = Select Tools`

This isnt mandatory for this PR as far as I'm concerned but it seems a shame not to do it at this stage as theres only a few instances right now?

**Comment by zersiax** [comment]
Agreed. Will take some time today to fix the linting errors and add the i18n support. Do we have access to $t() within inject.js for the frame titles, as well?

**Comment by psiinon** [comment]
Possibly not :/ You could try and see - if not we'll have to look into that...

**Comment by zersiax** [comment]
@psiinon  doesn't look like it :( getting $t undefined errors when I try. I will leave those entries alone for now, then, can certainly do the other ones, though. Give me half an hour.

**Comment by zersiax** [comment]
@psiinon  if I add the $t() calls in the aria-label attributes of the elements like yousuggessted, it seems they don't get evaluated properly, or not in time. Example:
			<hud-button icon="<<ZAP_HUD_FILES>>/image/plus.png" name="add-tool" aria-label="$t('message.a11y_label_add_tool')"></hud-button>


This results in the label getting set to that hardcoded string , rather than it getting evaluated. I also tried by doing "{{ $t('message.a11y_label_add_tool'}}"wih similar results.
I also see I didn't seem to have gotten to all the linting errors :) checking

File: src/main/zapHomeFiles/hud/display.html [hunk 1]
```diff
@@ -108,7 +108,7 @@
 
         <!-- select tool modal component -->
         <template id="select-tool-modal-template">
-            <modal :title="title" :show="show" @close="close">
+            <modal :title="title" :show="show" @close="close" aria-label="Select Tools">
```
Comment (psiinon) [inline src/main/zapHomeFiles/hud/display.html:111]: Worth i18n'ing this and the other ones in this file?

File: src/main/zapHomeFiles/hud/panel.html [hunk 2]
```diff
@@ -15,13 +15,13 @@
 
 	<template id="hud-buttons-template">
 		<div id="hud-buttons" :class="{'d-hide': !isVisible, 'd-visible': isVisible}">
-			<hud-button v-for="tool in tools" :key="tool.name" :label="tool.label" :icon="'<<ZAP_HUD_FILES>>/image/' + tool.icon" :data="tool.data" :name="tool.name"></hud-button>
-			<hud-button icon="<<ZAP_HUD_FILES>>/image/plus.png" name="add-tool"></hud-button>
+			<hud-button v-for="tool in tools" :key="tool.name" :label="tool.label" :icon="'<<ZAP_HUD_FILES>>/image/' + tool.icon" :data="tool.data" :name="tool.name" role="button" :aria-label="label"></hud-button>
+			<hud-button icon="<<ZAP_HUD_FILES>>/image/plus.png" name="add-tool" aria-label="Add Tool"></hud-button>
```
Comment (psiinon) [inline src/main/zapHomeFiles/hud/panel.html:19]: Hopefully these can be i18n'ed as well...

**Comment by zersiax** [comment]
@psiinon  I wrote strings for all of those, but like I stated in the comment above the translations don't appear to be taking :) Could you double-check the syntax you gave me for correctness? If it is correct, we might be running into a limitation of the way ARIA properties on elements are calculated by the browser; JS might be coming in too late to influence the way they are rendered. 
For completeness, if I set the  "add tools" to "$t('message.a11y_label_add_tools')" insead on the aria-label attribute, the screen reader will literally read out $t('message.a11y_label_add_tools) rather than the retrieved value, and I'm not sure why that is happening. I tried variations with adding and removing a : in front of the aria-label property, as well as using {{}} within the double quotes to try to explicitly interpolate the JS, but it doesn't appear to be having an effect in my tests.

**Comment by psiinon** [comment]
Arg, sorry! I'll have a look at that tomorrow:)

**Comment by psiinon** [comment]
Based on this example: https://vuejsdevelopers.com/2019/05/13/accessibility-validation-aria-vuejs/
I've worked out that putting a `:` immediatelly before `aria-label=$t('message.a11y_label_close')` makes it work!
So `:aria-label=$t('message.a11y_label_close')` :)
OK, so I dont really know why thats needed, but it makes it work for me :D

**Comment by zersiax** [comment]
@psiinon confirmed working, I'll make the requisite changes. I also noticed that panel.html wasn't i18n-enabled, I've fixed that.
I'm also noticing zap v2.1.0 is being used by gradle, need me to bump that up to 2.1.1 while I'm here?

**Comment by psiinon** [comment]
2.11.1? :) please!

**Comment by thc202** [comment]
Update the changelog as well :)

File: src/main/zapHomeFiles/hud/drawer.html [hunk 3]
```diff
@@ -140,7 +140,7 @@
                         <drawer-button-settings></drawer-button-settings>
                     </li>
                     <li v-if="tabsVisible" class="tab-item tab-action">
-                        <button class="btn btn-action btn-sm"  @click="toggleOpenClose">
+                        <button class="btn btn-action btn-sm"  @click="toggleOpenClose" aria-label="Open/Close">
```
Comment (psiinon) [inline src/main/zapHomeFiles/hud/drawer.html:143]: i18n as well?

File: src/main/zapHomeFiles/hud/drawer.html [hunk 4]
```diff
@@ -154,18 +154,18 @@
...
-        <div :class="{'hud-button': true, 'small': true, 'active': isActive}" @click="$emit('click');" @mouseover="mouseOver" @mouseleave="mouseLeave">
+        <div :class="{'hud-button': true, 'small': true, 'active': isActive}" @click="$emit('click');" @mouseover="mouseOver" @mouseleave="mouseLeave" role="button">
             <img :src="icon" />
             <span>{{label}}</span>
         </div>
     </template>
 
     <template id="drawer-button-settings-template">
-        <drawer-button-template :icon="icon" @click="showHudSettings"></drawer-button-template>
+        <drawer-button-template :icon="icon" @click="showHudSettings" aria-label="Settings"></drawer-button-template>
```
Comment (psiinon) [inline src/main/zapHomeFiles/hud/drawer.html:164]: Ditto ;)

File: src/main/zapHomeFiles/hud/drawer.html [hunk 5]
```diff
@@ -154,18 +154,18 @@
...
     </template>
 
     <template id="drawer-button-showhide-template">
-        <drawer-button-template :icon="icon" @click="toggleIsVisible"></drawer-button-template>
+        <drawer-button-template :icon="icon" @click="toggleIsVisible" aria-label="Show/Hide HUD"></drawer-button-template>
```
Comment (psiinon) [inline src/main/zapHomeFiles/hud/drawer.html:168]: Ditto ;)

**Review by psiinon — COMMENTED** [review]
A couple more places that can be internationalised - shame not to do that now?
But otherwise looking good!

**Comment by zersiax** [comment]
good catch, i forgot to put those three back. fixed now.

**Review by psiinon — APPROVED** [review]

**Comment by psiinon** [comment]
Many thanks!

**Comment by psiinon** [comment]
@thc202 ok with this PR?

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,7 +6,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-
+- Added accessibility fixes for rapid navigation between panels. 
+- added semantic roles and better labeling for screen reader users
+- Bumped ZAP version in build.gradle.kts to v2.11.1
```
Comment (thc202) [inline CHANGELOG.md:11]:
```suggestion
### Added
- Added accessibility fixes for rapid navigation between panels. 
- Added semantic roles and better labelling for screen reader users.

### Changed
- Update minimum ZAP version to 2.11.1.

```

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,7 +6,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 ## [Unreleased]
 
+#### Added
 
+- Added accessibility fixes for rapid navigation between panels. 
+- added semantic roles and better labeling for screen reader users
+
+#### Changed
+
+- Bumped ZAP version in build.gradle.kts to v2.11.1
```
Comment (thc202) [inline CHANGELOG.md:16]:
```suggestion
### Added
- Added accessibility fixes for rapid navigation between panels. 
- Added semantic roles and better labelling for screen reader users.
### Changed
- Update minimum ZAP version to 2.11.1.
```

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**